llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -3095,6 +3096,23 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
         MaybeSimplifyHint(OBU.Inputs[0]);
         MaybeSimplifyHint(OBU.Inputs[1]);
       }
+
+      // Try to fold alignment assumption into a load's !align metadata, if the
+      // assumption is valid in the load's context.
+      if (OBU.getTagName() == "align" && OBU.Inputs.size() == 2) {
+        auto *LI = dyn_cast<LoadInst>(OBU.Inputs[0]);
+        if (!LI ||
+            !isValidAssumeForContext(II, LI, &DT, /*AllowEphemerals=*/true))
+          continue;
+        auto *Align = cast<ConstantInt>(OBU.Inputs[1]);
+        if (!isPowerOf2_64(Align->getZExtValue()))
```

Comment:
```suggestion
        auto *Align = dyn_cast<ConstantInt>(OBU.Inputs[1]);
        if (!Align || !isPowerOf2_64(Align->getZExtValue()))
```
> While attributes expect constant arguments, assume operand bundles may be provided a dynamic value, for example:
```
call void @llvm.assume(i1 true) ["align"(ptr %val, i32 %align)]
```

Reply:
Can we use `getKnowledgeFromBundle` here?

Reply:
Yes thanks! Didn't check that they also allow dynamic values.

Reply:
(also added a check of the type size + test)

Reply:
> Can we use getKnowledgeFromBundle here?

Missed this earlier, updated now, thanks!